File: projects/neors232/java_NeoRS232/neoRS232/src/SimpleWrite.java
/*!\file SimpleWrite.java
 * Contains low-level functions for writing data to the 
 * serial port.
 */
 
package LieBTrau.comm;

import java.io.*;
import java.util.*;
import javax.comm.*;

/*!Class that will write the data to the serial port
 */
class SimpleWrite implements Protocol {
	
    private OutputStream outputStream;
	
	/*!Constructor
	 *\param serialPort Handler to the serialPort
	 */
	SimpleWrite(SerialPort serialPort){
    	try {
        	outputStream = serialPort.getOutputStream();
        } catch (IOException e) {
        	System.err.println(e);
        }
	}//constructor
	
	/*!Check if the received character is a flag character.
	 *\param c The received data byte.
	 */
	boolean vlagKarakterInData(byte c){
		if((c==LEADING_FLAG||
			c==TRAILING_FLAG)||
			(c==DLE_FLAG||
			c==ACK0_FLAG))return true;
		else return false;
	}//vlagKarakter
	
	/*!Write a dataframe to the serial port
	 *\param buffer Buffer containing the data
	 *\param typeFrame Type of the frame (ordinary or ack).
	 */
	void schrijfString(byte[] buffer, int typeFrame){
		/*Er is nog geen beperking in de buffergrootte, een grote
		 *hoeveelheid data kan de ontvangstbuffer van de ontvanger 
		 *overspoelen
		 */
		try {
			switch(typeFrame){
				
				case INFO_FRAME:
					//Leading flag
					outputStream.write(LEADING_FLAG);
					//Data	
					for(int i=0;i<buffer.length;i++){		
        				if(vlagKarakterInData(buffer[i]))outputStream.write(DLE_FLAG);
        				outputStream.write(buffer[i]);
        			}
					//Trailing flag
		        	outputStream.write(TRAILING_FLAG);
		        	break;
		        	
		        case ACK_FRAME:
		        	outputStream.write(ACK0_FLAG);
		        	break;
		        	
		        default:
		        	System.err.println("Geen type gegeven");
		        	
		   	}    	
        	outputStream.flush();
        } catch (IOException e) {
        	System.err.println("k1:" +e);
        }
    }//SchrijfString
    
    /*!Close outputstream, so no more data can be sent.
     */
    void close() throws IOException{
    	outputStream.flush();
    	outputStream.close();
    }//close
    
}//class

File: projects/neors232/AVR_NeoRS232/src/simpleWrite.c
/*!\file simpleWrite.c
 *Functions implementing the protocol: writing data
 */


#include "define.h"


BOOL flagCharacterInData(uint8_t c);


/*! Write a databuffer to the serial port.  No CRC will be appended.
 * This function adds the flags for the protocol and performs the 
 * bytestuffing.  Finally it sends the data to the UART.
 *\param buffer Buffer containing the data.
 *\param length Length of the databuffer
 *\param typeFrame Type of frame (info frame or acknowledge frame)
 */
void writeString(uint8_t* buffer, uint8_t length, FRAME_TYPE typeFrame){
		uint8_t i=0;
		switch(typeFrame){
			case INFO_FRAME:
				//Leading flag
				write_USART_B(LEADING_FLAG);
				/*Data:	
				 * if a databyte is equal to a flag character, then prepend
				 * a DLE_FLAG to it.
				 */
				for(i=0;i<length;i++){		
       				if(flagCharacterInData(buffer[i]))write_USART_B(DLE_FLAG);
       				write_USART_B(buffer[i]);
       			}
				//Trailing flag
	        	write_USART_B(TRAILING_FLAG);
	        	break;		        	
	        case ACK_FRAME:
	        	write_USART_B(ACK0_FLAG);
	        	break;		        	
	   	}    	
}//writeString


/*!Check if the character is a flag character.
 *\param c The character
 *\return TRUE when c is a flag character else FALSE.
 */
BOOL flagCharacterInData(uint8_t c){
	if(c==LEADING_FLAG||
		c==TRAILING_FLAG||
		c==DLE_FLAG||
		c==ACK0_FLAG)return TRUE;
	else return FALSE;
}//flagCharacterInData

